File: 3363-find-the-maximum-number-of-fruits-collected/3363-find-the-maximum-number-of-fruits-collected.java
class Solution {
    Integer[][] dp;
    public Map<Integer,List<List<Integer>>> mp = Map.of(
        0,List.of(List.of(1,1),List.of(1,-1),List.of(1,0)),
        1,List.of(List.of(0,1),List.of(-1,1),List.of(1,1))
    );
    int n;
    public int maxCollectedFruits(int[][] fruits) {
        n = fruits.length;
        dp = new Integer[n][n]; 
        int ans = solve(0,n-1,0,fruits) + solve(n-1,0,1,fruits);
        for(int i=0;i<n;i++) ans+=fruits[i][i];
        return ans;
    }

    boolean isValid(int row,int col) {
        return row>=0 && row<n && col>=0 && col<n;
    }

    int solve(int row,int col,int child,int[][] fruits) {
        if(row==n-1 && col==n-1) return 0;
        if((child==0 && col<=row) || (child==1 && row<=col)) return Integer.MIN_VALUE;
        if (dp[row][col] != null) return dp[row][col];

        int ans = 0;
        List<List<Integer>> dir = mp.get(child);
        for(int i=0;i<dir.size();i++) {
            int newRow = row + dir.get(i).get(0);
            int newCol = col + dir.get(i).get(1);
            if(!isValid(newRow,newCol)) continue;
            ans = Math.max(ans , fruits[row][col] + solve(newRow,newCol,child,fruits));
        }
        return dp[row][col]=ans;
    }
}

File: 3363-find-the-maximum-number-of-fruits-collected/3363-find-the-maximum-number-of-fruits-collected.cpp
class Solution {
public:
    int maxCollectedFruits(vector<vector<int>>& fruits) {
        int diagonal = 0;
        int n= fruits.size();
        for(int i=0;i<n;i++){
            diagonal+= fruits[i][i];
        }     

        vector<vector<int>> dp(n+1,vector<int>(n+1,-1));
        dp[0][n-1] = fruits[0][n-1];
        for(int i=1;i<n;i++) {
            for(int j = n-1; j>i ; j--) {
                int maxVal = max({dp[i-1][j-1], dp[i-1][j], dp[i-1][j+1]});
                if(maxVal!=-1){
                    dp[i][j] = fruits[i][j] + maxVal;
                }
            }
        }
        int i=n-1,j=n-1;
        dp[i][j] = max({dp[i-1][j-1], dp[i-1][j], dp[i-1][j+1]});

        vector<vector<int>> dp1(n+1,vector<int>(n+1,-1));
        dp1[n-1][0] = fruits[n-1][0];
        for(int j=1;j<n;j++) {
            for(int i = n-1; i>j ; i--) {
                int maxVal = max({dp1[i-1][j-1], dp1[i][j-1], dp1[i+1][j-1]});
                if(maxVal!=-1){
                    dp1[i][j] = fruits[i][j] + maxVal;
                }
            }
        }
        dp1[i][j] = max({dp1[i-1][j-1], dp1[i][j-1], dp1[i+1][j-1]});

        return diagonal + dp[n-1][n-1] + dp1[n-1][n-1];
    }
};


/*

1. Sum up the main diagonal

Formula
1. bottom left: dp[i][j]=fruits[i][j] +max(dp[i-1][j-1], dp[i][j-1], dp[i+1][j-1])
2. top right: dp[i][j]=fruits[i][j] + max(dp[i-1][j-1], dp[i-1][j], dp[i-1][j+1])



*/